src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs

```diff
@@ -20,35 +20,108 @@ internal sealed class ModelTypeProvider : TypeProvider
         protected override string DefaultName { get; }
         protected override string DefaultAccessibility { get; }
 
-        public IReadOnlyList<FieldDeclaration> Fields { get; }
-        public ConstructorSignature PublicConstructor { get; }
-        public ConstructorSignature SerializationConstructor { get; }
+        // adopt lazy initialization for fields related properties
+        public IReadOnlyList<FieldDeclaration> Fields
+        {
+            get
+            {
+                if (_fields != null)
+                {
+                    return _fields;
+                }
+
+                _fieldsInitializer();
+                return _fields!;
+            }
+        }
+        public ConstructorSignature PublicConstructor
+        {
+            get
+            {
+                if (_publicConstructor != null)
+                {
+                    return _publicConstructor;
+                }
+
+                _fieldsInitializer();
+                return _publicConstructor!;
+            }
+        }
+        public ConstructorSignature SerializationConstructor
+        {
+            get
+            {
+                if (_serializationConstructor != null)
+                {
+                    return _serializationConstructor;
+                }
+
+                _fieldsInitializer();
+                return _serializationConstructor!;
+            }
+        }
+
+        private IReadOnlyDictionary<FieldDeclaration, InputModelProperty> FieldsToInputs
+        {
+            get
+            {
+                if (_fieldsToInputs != null)
+                {
+                    return _fieldsToInputs;
+                }
 
-        private readonly IReadOnlyDictionary<FieldDeclaration, InputModelProperty> _fieldsToInputs;
+                _fieldsInitializer();
+                return _fieldsToInputs!;
+            }
+        }
         // parameter name should be unique since it's bound to field property
-        private readonly IReadOnlyDictionary<string, FieldDeclaration> _parameterNamesToFields;
+        private IReadOnlyDictionary<string, FieldDeclaration> ParameterNamesToFields
+        {
+            get
+            {
+                if (_parameterNamesToFields != null)
+                {
+                    return _parameterNamesToFields;
+                }
+
+                _fieldsInitializer();
+                return _parameterNamesToFields!;
+            }
+        }
+
+        // lazy envaluation
+        private readonly Action _fieldsInitializer;
```

Comment:
is it better to name it _modelInitializer?

Reply:
The `ModelTypeProvider` is partially useful after the constructor is invoked. What remains lazy evaluation is the fields related properties. So I name it after `_fieldsInitializer`.